File: backend/extract.ts
#!/usr/bin/env tsx
/**
 * Extraction step: email_id → context → LLM → extraction_runs row.
 *
 * Entry points:
 *   CLI:    tsx backend/extract.ts <email_uuid>
 *   Server: tsx backend/extract.ts --server
 */

import { createServer } from 'http'
import fs from 'fs/promises'
import path from 'path'
import { fileURLToPath } from 'url'

import OpenAI from 'openai'
import pg from 'pg'
import { z } from 'zod'
import { PDFParse } from 'pdf-parse'
import mammoth from 'mammoth'
import XLSX from 'xlsx'

import { SYSTEM_PROMPT, formatContext } from './prompt.js'

const { Pool } = pg
const pool = new Pool({ connectionString: process.env.DATABASE_URL })
const MODEL_NAME = process.env.MODEL_NAME ?? 'anthropic/claude-sonnet-4'

// ── Zod schema ──────────────────────────────────────────────────────────────

const LineUpdateSchema = z.object({
  sku_or_code: z.string(),
  field: z.enum(['delivery_date', 'quantity']),
  new_value: z.string(),
  evidence: z.string(),
  confidence: z.number().min(0).max(1),
})

const ExtractionSchema = z.object({
  po_updates: z
    .array(
      z.object({
        po_ref: z.string(),
        source: z.string().optional().default('body'),
        evidence: z.string(),
        confidence: z.number().min(0).max(1),
        line_updates: z.array(LineUpdateSchema).default([]),
      }),
    )
    .default([]),
  unmatched_mentions: z.array(z.string()).default([]),
})

function parseModelContent(content: string): unknown {
  const stripped = content.trim()
  // Strip markdown code fences if present
  const fenced = stripped.match(/```(?:json)?\s*([\s\S]*?)\s*```/)
  try {
    return JSON.parse(fenced ? fenced[1] : stripped)
  } catch {
    return {}
  }
}

// ── context loading ─────────────────────────────────────────────────────────

async function loadContext(senderRaw: string) {
  const match = senderRaw.match(/<([^>]+)>/)
  const senderEmail = (match ? match[1] : senderRaw).toLowerCase()

  // Check supplier.email first, then fall back to supplier_email_aliases
  const { rows: supplierRows } = await pool.query<{ id: string; name: string }>(
    `SELECT s.id, s.name FROM supplier s WHERE lower(s.email) = $1
     UNION
     SELECT s.id, s.name FROM supplier s
     JOIN supplier_email_aliases sea ON sea.supplier_id = s.id
     WHERE lower(sea.email_address) = $1
     LIMIT 1`,
    [senderEmail],
  )
  const supplier = supplierRows[0] ?? null
  if (!supplier) return { supplier: null, pos: [], aliases: [] }

  const { rows: poRows } = await pool.query(
    `SELECT po.reference_num,
            po.delivery_date     AS po_delivery_date,
            pol.quantity,
            pol.delivery_date    AS line_delivery_date,
            p.sku,
            p.title              AS product_name
     FROM   purchase_order po
     JOIN   purchase_order_line pol ON pol.purchase_order_id = po.id
     JOIN   product p               ON p.id = pol.product_id
     WHERE  po.supplier_id = $1
     ORDER  BY po.reference_num`,
    [supplier.id],
  )

  const { rows: aliasRows } = await pool.query(
    `SELECT sp.supplier_sku, p.sku, p.title AS product_name
     FROM   supplier_product sp
     JOIN   product p ON p.id = sp.product_id
     WHERE  sp.supplier_id = $1 AND sp.supplier_sku IS NOT NULL`,
    [supplier.id],
  )

  const pos = poRows.map((r) => ({
    po_ref:            r.reference_num,
    po_delivery_date:  r.po_delivery_date ? String(r.po_delivery_date) : null,
    quantity:          String(r.quantity),
    line_delivery_date: r.line_delivery_date ? String(r.line_delivery_date) : null,
    sku:               r.sku,
    product_name:      r.product_name,
  }))

  const aliases = aliasRows.map((r) => ({
    supplier_sku: r.supplier_sku,
    sku:          r.sku,
    product_name: r.product_name,
  }))

  return { supplier, pos, aliases }
}

// ── LLM client ──────────────────────────────────────────────────────────────

const client = new OpenAI({
  baseURL: 'https://openrouter.ai/api/v1',
  apiKey: process.env.OPENROUTER_API_KEY,
})

// ── attachment loading ───────────────────────────────────────────────────────

const ATTACHMENTS_DIR = process.env.ATTACHMENTS_DIR
  ? path.resolve(process.env.ATTACHMENTS_DIR)
  : path.resolve('attachments')

interface ImageAttachment {
  originalName: string
  mimeType: string
  base64: string
}

interface TextAttachment {
  originalName: string
  mimeType: string
  text: string
}

// ── document text extraction ─────────────────────────────────────────────────

const XLSX_MIME_TYPES = new Set([
  'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
  'application/vnd.ms-excel',
  'application/x-excel',
])

const DOCX_MIME_TYPES = new Set([
  'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
  'application/msword',
])

async function extractDocumentText(
  data: Buffer,
  mimeType: string,
  originalName: string,
): Promise<string | null> {
  try {
    if (mimeType === 'application/pdf') {
      const parser = new PDFParse({ data })
      const result = await parser.getText()
      const text = result.text.trim()
      if (text) return text
      // No text layer — likely a scanned image PDF
      return `[Scanned PDF — no text layer detected. A reviewer should inspect "${originalName}" directly.]`
    }

    if (XLSX_MIME_TYPES.has(mimeType)) {
      const wb = XLSX.read(data, { type: 'buffer' })
      return wb.SheetNames.map((name) => {
        const ws = wb.Sheets[name]
        return `[Sheet: ${name}]\n${XLSX.utils.sheet_to_csv(ws)}`
      }).join('\n\n')
    }

    if (DOCX_MIME_TYPES.has(mimeType)) {
      const result = await mammoth.extractRawText({ buffer: data })
      return result.value.trim() || null
    }

    if (mimeType === 'text/csv' || mimeType === 'text/plain') {
      return data.toString('utf-8').trim() || null
    }

    return null // unsupported format — silently skip
  } catch {
    return null
  }
}

// ── attachment loading ────────────────────────────────────────────────────────

interface AttachmentResult {
  images: ImageAttachment[]
  texts: TextAttachment[]
}

async function loadAttachments(emailId: string): Promise<AttachmentResult> {
  const { rows } = await pool.query<{
    stored_name: string
    original_name: string
    mime_type: string
  }>(
    `SELECT stored_name, original_name, mime_type
     FROM email_attachments
     WHERE email_id = $1`,
    [emailId],
  )

  const images: ImageAttachment[] = []
  const texts: TextAttachment[] = []

  for (const row of rows) {
    let data: Buffer
    try {
      data = await fs.readFile(path.join(ATTACHMENTS_DIR, row.stored_name))
    } catch {
      continue // skip unreadable files
    }

    if (row.mime_type.startsWith('image/')) {
      images.push({
        originalName: row.original_name,
        mimeType: row.mime_type,
        base64: data.toString('base64'),
      })
    } else {
      const text = await extractDocumentText(data, row.mime_type, row.original_name)
      if (text !== null) {
        texts.push({ originalName: row.original_name, mimeType: row.mime_type, text })
      }
    }
  }

  return { images, texts }
}

// ── core extraction ─────────────────────────────────────────────────────────

export async function extract(emailId: string): Promise<string> {
  const { rows } = await pool.query(
    'SELECT sender, subject, body_text FROM emails WHERE id = $1',
    [emailId],
  )
  if (!rows.length) throw new Error(`email not found: ${emailId}`)

  const { sender, subject, body_text: bodyText } = rows[0]
  const [{ supplier, pos, aliases }, { images: imageAttachments, texts: textAttachments }] =
    await Promise.all([loadContext(sender), loadAttachments(emailId)])

  // Base text: context + email body + all extractable document text
  const baseText = [
    formatContext(supplier, pos, aliases),
    `## Email\nFrom: ${sender}\nSubject: ${subject}\n\n${bodyText ?? '(no body)'}`,
    ...textAttachments.map((t) => `## Attachment: ${t.originalName}\n${t.text}`),
  ].join('\n\n')

  type ContentPart =
    | { type: 'text'; text: string }
    | { type: 'image_url'; image_url: { url: string } }

  const userContent: string | ContentPart[] =
    imageAttachments.length === 0
      ? baseText
      : [
          { type: 'text', text: baseText },
          ...imageAttachments.map((img): ContentPart => ({
            type: 'image_url',
            image_url: { url: `data:${img.mimeType};base64,${img.base64}` },
          })),
        ]

  let llmOutput: object = {}
  let status: 'success' | 'error' = 'success'
  let errorMessage: string | null = null

  const callLLM = async (content: string | ContentPart[]) => {
    try {
      const response = await client.chat.completions.create({
        model: MODEL_NAME,
        messages: [
          { role: 'system', content: SYSTEM_PROMPT },
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          { role: 'user', content: content as any },
        ],
      })
      return response.choices[0].message.content ?? ''
    } catch (err) {
      // Log full OpenRouter error body for debugging
      if (err instanceof OpenAI.APIError) {
        console.error(
          `[extract] OpenRouter error ${err.status} (model=${MODEL_NAME}):`,
          JSON.stringify(err.error ?? err.message),
        )
        throw new Error(`OpenRouter ${err.status}: ${JSON.stringify(err.error ?? err.message)}`)
      }
      throw err
    }
  }

  try {
    let raw: string
    try {
      raw = await callLLM(userContent)
    } catch (visionErr) {
      // Retry text-only if the model doesn't support vision (404 or 500)
      const msg = String(visionErr)
      if (Array.isArray(userContent) && (msg.includes('image') || msg.includes('500'))) {
        console.error('[extract] Vision call failed, retrying text-only:', msg)
        raw = await callLLM(baseText)
      } else {
        throw visionErr
      }
    }

    const parsed = ExtractionSchema.safeParse(parseModelContent(raw))
    if (!parsed.success) {
      throw new Error(`Schema validation failed: ${parsed.error.message}`)
    }
    llmOutput = parsed.data
  } catch (err) {
    status = 'error'
    errorMessage = String(err)
  }

  const db = await pool.connect()
  let runId = ''
  try {
    await db.query('BEGIN')

    const runResult = await db.query(
      `INSERT INTO extraction_runs (email_id, model_version, llm_output, status, error_message)
       VALUES ($1, $2, $3, $4, $5)
       RETURNING id`,
      [emailId, MODEL_NAME, JSON.stringify(llmOutput), status, errorMessage],
    )
    runId = String(runResult.rows[0].id)

    await db.query('UPDATE emails SET status = $1 WHERE id = $2', [
      status === 'success' ? 'extracted' : 'failed',
      emailId,
    ])

    await db.query('COMMIT')
  } catch (err) {
    await db.query('ROLLBACK').catch(() => {})
    throw err
  } finally {
    db.release()
  }

  if (status === 'error') throw new Error(`Extraction failed: ${errorMessage}`)
  return runId
}

// ── HTTP server ─────────────────────────────────────────────────────────────

function startServer() {
  createServer(async (req, res) => {
    if (req.method !== 'POST' || req.url !== '/extract') {
      res.writeHead(404).end()
      return
    }
    const body = await new Promise<string>((resolve) => {
      let data = ''
      req.on('data', (chunk) => { data += chunk })
      req.on('end', () => resolve(data))
    })
    try {
      const runId = await extract(body.trim())
      res.writeHead(200, { 'Content-Type': 'text/plain' }).end(runId)
    } catch (err) {
      res.writeHead(500).end(String(err))
    }
  }).listen(8001, () => console.log('extract server listening on :8001'))
}

// ── entrypoint ──────────────────────────────────────────────────────────────

if (fileURLToPath(import.meta.url) === process.argv[1]) {
  const arg = process.argv.slice(2).find((a) => a !== '--')
  if (arg === '--server') {
    startServer()
  } else if (arg) {
    extract(arg)
      .then(console.log)
      .catch((e) => { console.error(e); process.exit(1) })
  } else {
    console.error('usage: tsx backend/extract.ts <email_uuid> | --server')
    process.exit(1)
  }
}

File: backend/extract.py
"""Extraction step: email_id → context → LLM → extraction_runs row.

The email body, the supplier's known purchase orders, their product-code
aliases and their past corrections are assembled into one prompt. Whatever
the model returns is parsed and validated against `ExtractionOutput` before
it is written. A run is always recorded — on failure with status='error'
and the message — so a bad extraction is a stored fact rather than a crash.
"""

from __future__ import annotations

import base64
import json
import logging
import re
from pathlib import Path
from typing import Any

from psycopg import AsyncConnection
from pydantic import ValidationError

from backend.config import get_settings
from backend.db import connection
from backend.llm import Content, ExtractionProvider, build_provider
from backend.prompt import SYSTEM_PROMPT, format_context
from backend.schemas import ExtractionOutput
from backend.suppliers import resolve_supplier

log = logging.getLogger(__name__)

_FENCE = re.compile(r"```(?:json)?\s*([\s\S]*?)\s*```")


class ExtractionError(RuntimeError):
    """Raised after the failed run has been recorded."""


class EmailNotFoundError(LookupError):
    pass


def parse_model_content(content: str) -> Any:
    """Parse raw assistant text into JSON, tolerating markdown code fences.

    Returns `{}` when the text is not JSON at all, which validates as an
    empty extraction — see `ExtractionOutput`'s field defaults.
    """
    stripped = content.strip()
    fenced = _FENCE.search(stripped)
    try:
        return json.loads(fenced.group(1) if fenced else stripped)
    except ValueError:
        return {}


# ── context loading ──────────────────────────────────────────────────────────


async def load_context(
    conn: AsyncConnection,
    sender_raw: str,
) -> tuple[dict[str, Any] | None, list[Any], list[Any], list[Any]]:
    """Everything known about the sender, as the prompt needs it."""
    supplier = await resolve_supplier(conn, sender_raw)
    if supplier is None:
        return None, [], [], []

    cur = await conn.execute(
        """
        SELECT po.reference_num,
               po.delivery_date     AS po_delivery_date,
               pol.quantity,
               pol.delivery_date    AS line_delivery_date,
               p.sku,
               p.title              AS product_name
        FROM   purchase_order po
        JOIN   purchase_order_line pol ON pol.purchase_order_id = po.id
        JOIN   product p               ON p.id = pol.product_id
        WHERE  po.supplier_id = %s
        ORDER  BY po.reference_num
        """,
        (supplier["id"],),
    )
    po_rows = await cur.fetchall()

    cur = await conn.execute(
        """
        SELECT sp.supplier_sku, p.sku, p.title AS product_name
        FROM   supplier_product sp
        JOIN   product p ON p.id = sp.product_id
        WHERE  sp.supplier_id = %s AND sp.supplier_sku IS NOT NULL
        """,
        (supplier["id"],),
    )
    alias_rows = await cur.fetchall()

    cur = await conn.execute(
        """
        SELECT context, wrong, correct, field
        FROM supplier_corrections
        WHERE supplier_id = %s
        ORDER BY created_at DESC
        LIMIT 5
        """,
        (supplier["id"],),
    )
    corrections = await cur.fetchall()

    pos = [
        {
            "po_ref": r["reference_num"],
            "po_delivery_date": r["po_delivery_date"].isoformat()
            if r["po_delivery_date"]
            else None,
            "quantity": str(r["quantity"]),
            "line_delivery_date": r["line_delivery_date"].isoformat()
            if r["line_delivery_date"]
            else None,
            "sku": r["sku"],
            "product_name": r["product_name"],
        }
        for r in po_rows
    ]

    aliases = [
        {
            "supplier_sku": r["supplier_sku"],
            "sku": r["sku"],
            "product_name": r["product_name"],
        }
        for r in alias_rows
    ]

    return supplier, pos, aliases, corrections


# ── attachment loading ───────────────────────────────────────────────────────


async def load_attachments(
    conn: AsyncConnection,
    email_id: str,
    attachments_dir: Path,
) -> tuple[list[dict[str, str]], list[dict[str, str]]]:
    """Split an email's attachments into vision images and extracted text."""
    from backend.attachments import extract_document_text

    cur = await conn.execute(
        """
        SELECT stored_name, original_name, mime_type
        FROM email_attachments
        WHERE email_id = %s
        """,
        (email_id,),
    )
    rows = await cur.fetchall()

    images: list[dict[str, str]] = []
    texts: list[dict[str, str]] = []

    for row in rows:
        try:
            data = (attachments_dir / row["stored_name"]).read_bytes()
        except OSError:
            continue  # skip unreadable files

        if row["mime_type"].startswith("image/"):
            images.append(
                {
                    "original_name": row["original_name"],
                    "mime_type": row["mime_type"],
                    "base64": base64.b64encode(data).decode("ascii"),
                }
            )
        else:
            text = extract_document_text(data, row["mime_type"], row["original_name"])
            if text is not None:
                texts.append(
                    {
                        "original_name": row["original_name"],
                        "mime_type": row["mime_type"],
                        "text": text,
                    }
                )

    return images, texts


# ── core extraction ──────────────────────────────────────────────────────────


async def extract(
    email_id: str,
    provider: ExtractionProvider | None = None,
) -> str:
    """Run extraction for one email. Returns the new extraction_runs id."""
    settings = get_settings()
    provider = provider or build_provider()

    async with connection() as conn:
        cur = await conn.execute(
            "SELECT sender, subject, body_text FROM emails WHERE id = %s",
            (email_id,),
        )
        email_row = await cur.fetchone()
        if email_row is None:
            raise EmailNotFoundError(f"email not found: {email_id}")

        sender = email_row["sender"]
        subject = email_row["subject"]
        body_text = email_row["body_text"]

        supplier, pos, aliases, corrections = await load_context(conn, sender)
        images, texts = await load_attachments(
            conn, email_id, settings.attachments_dir
        )

    # Base text: context + email body + all extractable document text
    base_text = "\n\n".join(
        [
            format_context(supplier, pos, aliases, corrections),
            f"## Email\nFrom: {sender}\nSubject: {subject}\n\n{body_text or '(no body)'}",
            *(f"## Attachment: {t['original_name']}\n{t['text']}" for t in texts),
        ]
    )

    user_content: Content = base_text
    if images:
        user_content = [
            {"type": "text", "text": base_text},
            *(
                {
                    "type": "image_url",
                    "image_url": {
                        "url": f"data:{img['mime_type']};base64,{img['base64']}"
                    },
                }
                for img in images
            ),
        ]

    llm_output: dict[str, Any] = {}
    status = "success"
    error_message: str | None = None
    # Falls back to the configured provider's name if the call never returns,
    # so a failed run still records what was tried.
    model_version = getattr(provider, "name", "unknown")

    try:
        try:
            completion = await provider.complete(SYSTEM_PROMPT, user_content)
        except Exception as vision_err:
            # Retry text-only if the model doesn't support vision (404 or 500)
            message = str(vision_err)
            if isinstance(user_content, list) and (
                "image" in message or "500" in message
            ):
                log.error("Vision call failed, retrying text-only: %s", message)
                completion = await provider.complete(SYSTEM_PROMPT, base_text)
            else:
                raise

        # Whatever answered — the primary or something down the fallback
        # chain — is what gets recorded against the run.
        model_version = completion.model_version

        try:
            parsed = ExtractionOutput.model_validate(
                parse_model_content(completion.text)
            )
        except ValidationError as err:
            raise ValueError(f"Schema validation failed: {err}") from err
        llm_output = parsed.model_dump()
    except Exception as err:
        status = "error"
        error_message = str(err)

    async with connection() as conn:
        async with conn.transaction():
            cur = await conn.execute(
                """
                INSERT INTO extraction_runs
                  (email_id, model_version, llm_output, status, error_message)
                VALUES (%s, %s, %s, %s, %s)
                RETURNING id
                """,
                (
                    email_id,
                    model_version,
                    json.dumps(llm_output),
                    status,
                    error_message,
                ),
            )
            run_row = await cur.fetchone()
            assert run_row is not None
            run_id = str(run_row["id"])

            await conn.execute(
                "UPDATE emails SET status = %s WHERE id = %s",
                ("extracted" if status == "success" else "failed", email_id),
            )

    if status == "error":
        raise ExtractionError(f"Extraction failed: {error_message}")
    return run_id
